### ISSUE TITLE
gradient layout contract

### ISSUE THREAD

**Issue by credoxfromcqut** [pytorch/pytorch#58979]
## 🐛 Bug

# **Warning: grad and param do not obey the gradient layout contract. This is not an error, but may impair performance.**

when the warnning appear after a valid function, the gpu occupation 70% goes to 100%.and cpu ocupation 50% goes to 10%,
the valid function just do a forward pass and it will change the model's mode(train or eval) and recover the original mode.then ,after the loss computation done, loss.backward() excute ,the warnning shows.besides, the loss computation costs much time after this
```
dist_ap, relative_p_inds = torch.max(
    dist_mat[is_pos].contiguous().view(N, -1), 1, keepdim=True)  # costly
```
when annotate the valid function， no warnning, the  cpu and gpu  occupation keeps almost constant.
after my step-wise test on the code above, **dist_mat[is_pos]** takes much time,the key is that the size of it is fixed after and before the valid function...,so i have to ensure this dont bring error result.
`=========================================================`
[previous issue](https://github.com/pytorch/pytorch/issues/57706)

now i have found the solution for my code!!
and dont know how it works. hope someone can help me to understand why!!

here is my experience:

i use a numpy random matrix to replace the picture array, it works, no perf decrease.

then i add *.cpoy() to the img. works!!

``` 
    # optimizer.zero_grad(set_to_none=True)  this doesn't work, mentioned in previous issue.
    ims = Variable(self.TVT(torch.from_numpy(ims.copy()).float()))
    feat = self.model(ims)
    feat = feat.data.cpu().numpy()
```

i dont know why the img will impact the gradient contract?

not to mention that i have transfer the img to cuda, and the gradient also locate on GPU, so why would this happen from the view of you? 

i prefer to name it as a bug, intuitively there is no any realation between gradient (gpu) and the img (cpu). 

## To Reproduce
[here is the train source](https://github.com/huanghoujing/person-reid-triplet-loss-baseline/blob/a1c2bd20180cfaf225782717aeaaab461798abec/script/experiment/train.py)

## Environment
 - PyTorch Version (e.g., 1.0):1.7.0
 - OS (e.g., Linux):windows
 - How you installed PyTorch (`conda`, `pip`, source):--
 - Build command you used (if compiling from source):
 - Python version:3.6
 - CUDA/cuDNN version:
 - GPU models and configuration:1660super
 - Any other relevant information:

**Comment by ngimel** [comment]
Please provide a reproducible code example clearly identifying an error or perf problem you are seeing.

**Comment by quang-truong** [comment]
Strangely `optimizer.zero_grad(set_to_none=True)` helps my situation.
